proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,264 @@
+      [2 x %"dxbc::RootSignature::StaticSamplerDesc"]
+         [%"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 85, i32 1, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+             i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 1, 
+             i32 0, i32 0 }, 
+          %"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 21, i32 3, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+           i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 2, 
+           i32 0, i32 0 }], 
+      %1 { 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          { i32 0, i32 1, i32 8 }, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %2 { %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+        { i32 3, i32 244 }, 
+        [3 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+         [%"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 2, i32 1, i32 1, i32 0, i32 0, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 0, i32 8, i32 1, i32 0, i32 1, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 1, i32 -1, i32 1, i32 0, i32 1, i32 -1 }] 
+       }, 
+       %3 {
+           %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+           { i32 1, i32 324 }, 
+           [1 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+           [%"dxbc::RootSignature::ContainerDescriptorRange" 
+             { i32 3, i32 4, i32 0, i32 1, i32 0, i32 -1 }] }, 
+           %"dxbc::RootSignature::ContainerRootConstants" 
+             { i32 10, i32 0, i32 3 } 
+       }
+      }
+    }
+
+```
+
+* LLVM IR for root signature saved in metadata.
+
+```
+
+  ; named metadata for entry root signature
+  !hlsl.entry.rootsignatures = !{!3}
+  …
+  !3 = !{ptr @main, !4}
+  !4 = !{i32 2, i32 3, !5, !18}      ; version 1.1, flag 3, root params, statis samplers.
+  !5 = !{!6, !7, !8, !9, !14, !17}  ; root parameters.
+  !6 = !{i32 2, i32 0, i32 0, i32 1, i32 8}
+  !7 = !{i32 3, i32 0, i32 0, i32 0, i32 0}
+  !8 = !{i32 4, i32 0, i32 0, i32 0, i32 0}
+  !9 = !{i32 0, i32 0, !10}  ; first table
+  !10 = !{!11, !12, !13} ; table ranges.
+  !11 = !{i32 2, i32 1, i32 1, i32 0, i32 0}
+  !12 = !{i32 0, i32 8, i32 1, i32 0, i32 1}
+  !13 = !{i32 1, i32 -1, i32 1, i32 0, i32 1}
+  !14 = !{i32 0, i32 0, !15}  ; second table
+  !15 = !{!16}  ; table ranges.
+  !16 = !{i32 3, i32 4, i32 0, i32 1, i32 0}
+  !17 = !{i32 1, i32 0, i32 10, i32 0, i32 3}
+  !18 = !{!19, !20} ; static samplers.
+  !19 = !{i32 85, i32 1, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 1, i32 0, i32 0}
+  !20 = !{i32 21, i32 3, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 2, i32 0, i32 0}
```

Comment:
IMO, this is a huge step in the right direction. I suspect there is some opportunity for further refinement, but this is _much_ simpler to read and specify in the IR.

Can you document in more detail what each value in the metadata corresponds to? Even trying to map this to the structures from the structure-based approach is extremely tedious. I’d really like to see something like how [DXIL.rst](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#shader-properties-and-capabilities) documents its metadata structures.

Reply:
I believe this comment has still not been addressed.

Reply:
Oh, I see in your update you mixed documenting the schema with the example. I think we should separate the documentation of the schema from the example. The example is helpful for driving conversation, but is something we should not include in the technical documentation.

Reply:
Added separate section for metadata format.